### PR TITLE
Pass Cloudfront distribution id from serverless-next.js component to cloudfront component

### PR DESCRIPTION
**Pull request by pradyumna1**
This PR completes #477 and #209

### CODE REVIEW

**Comment by pradyumna1** [comment]
Hey @danielcondemarin, I need help in understanding what else should be added in this PR?

**Comment by danielcondemarin** [comment]
> Hey @danielcondemarin, I need help in understanding what else should be added in this PR?

I documented the proposal in https://github.com/serverless-nextjs/serverless-next.js/issues/477. 

I think we may be able to merge your PR and publish in alpha. Have you tested it yourself?

**Comment by pradyumna1** [comment]
Yes, I have tested the part where we can send the distribution id for an existing CloudFront distribution. I haven't tested the root path passing though. I think we can merge this one in alpha and work on setting the root path separately. I have updated the main README.md as well accordingly. :)

File: README.md [hunk 1]
```diff
@@ -126,7 +126,7 @@ myNextApplication:
 
 ### Custom CloudFront configuration
 
-To specify your own CloudFront inputs, just add any [aws-cloudfront inputs](https://github.com/serverless-components/aws-cloudfront#3-configure) under `cloudfront`:
+To specify your own CloudFront inputs, just add any [aws-cloudfront inputs](https://github.com/serverless-nextjs/serverless-next.js/tree/master/packages/serverless-components/aws-cloudfront#3-configure) under `cloudfront`:
```
Comment (pradyumna1) [inline README.md:129]: Earlier this was pointing to the original serverless cloudfront component.

File: README.md [hunk 2]
```diff
@@ -135,6 +135,8 @@ myNextApplication:
   component: serverless-next.js
   inputs:
     cloudfront:
+      # if you want to use an existing cloudfront distribution, provide it here
```
Comment (pradyumna1) [inline README.md:138]: should this be further improved?

**Comment by pradyumna1** [comment]
Setting PR to WIP. What would be the ideal flow if the provided distribution ID doesn't exist? Is an error message fine or should we print something more verbose? We can create another distribution and ask the user to add it to serverless.yml if he/she wants to handle multiple environments. Need guidance @danielcondemarin. @andrewgadziksonos.

**Comment by danielcondemarin** [comment]
> Setting PR to WIP. What should be the ideal flow if the provided distribution ID doesn't exist? Is an error message fine or should we print something more verbose? We can create another distribution and ask the user to add it to serverless.yml if he/she wants to handle multiple environments. Need guidance @danielcondemarin. @andrewgadziksonos.

I'd say just throw an error. If they want a new distribution just don't specify the distributionId.

File: README.md (hunk 2 above)
Comment (danielcondemarin) [reply README.md:138]: Yeah maybe provide a link to how "updating" existing distributions works. (https://github.com/serverless-nextjs/serverless-next.js/tree/master/packages/serverless-components/aws-cloudfront#updating-an-existing-cloudfront-distribution)

**Review by anatolii-albelli — APPROVED** [review]

**Comment by pradyumna1** [comment]
@danielcondemarin sorry I couldn't reply earlier. I have added another functionality. The distribution id will be printed in the serverless output so that the user can copy it easily.

**Review by benjipott — APPROVED** [review]

**Review by benjipott — DISMISSED** [review]

File: packages/serverless-components/nextjs-component/src/component.ts [hunk 3]
```diff
@@ -22,6 +22,7 @@ import type {
 type DeploymentResult = {
   appUrl: string;
   bucketName: string;
+  distributionId: unknown;
```
Comment (danielcondemarin) [inline packages/serverless-components/nextjs-component/src/component.ts:25]: Could this be made a string instead? 

```suggestion
  distributionId: string;
```

File: packages/serverless-components/nextjs-component/src/component.ts [hunk 4]
```diff
@@ -497,6 +499,7 @@ class NextjsComponent extends Component {
     delete defaultLambdaAtEdgeConfig["origin-response"];
 
     const cloudFrontOutputs = await cloudFront({
+      distributionId: cloudFrontDistributionId || null,
```
Comment (danielcondemarin) [inline packages/serverless-components/nextjs-component/src/component.ts:502]: Is the or to null necessary? 

```suggestion
      distributionId: cloudFrontDistributionId,
```

File: packages/serverless-components/nextjs-component/src/component.ts (hunk 3 above)
Comment (pradyumna1) [reply packages/serverless-components/nextjs-component/src/component.ts]: done

File: packages/serverless-components/nextjs-component/src/component.ts (hunk 4 above)
Comment (pradyumna1) [reply packages/serverless-components/nextjs-component/src/component.ts]: done

**Review by danielcondemarin — APPROVED** [review]

**Comment by michaelwindell26** [comment]
This looks perfect, please could we get a release including this @danielcondemarin ?

**Comment by vkhemlan** [comment]
We could really use this feature too!

Also the current documentation

https://github.com/serverless-nextjs/serverless-next.js

Mentions these new parameters that are not available on the latest release, so it was confusing for us that it didn't work.

**Comment by dphang** [comment]
Yeah, the docs are always for latest master, which may not be published yet. Thanks for the feedback, I am trying to think of ways to streamline the publishing process and/or make it easy to get the latest changes without needing to wait for @danielcondemarin or another collaborator to publish.

**Comment by dphang** [comment]
Anyway, I have published 1.17.0-alpha.10 which should have the latest changes.

Until there is a better solution, it is best to look for the specific release and browse the git tree to get the docs relevant for that: https://github.com/serverless-nextjs/serverless-next.js/releases. In the meanwhile, I will add a note in the docs to avoid confusion :)

**Comment by dphang** [comment]
Latest release has been published.

**Comment by danielcondemarin** [comment]
> Latest release has been published.

You should be able to use @sls-next/serverless-component@1.17.0-alpha.12  🙌

**Comment by michaelwindell26** [comment]
I can confirm using `distributionId` worked as expected using `@sls-next/serverless-component@1.17.0-alpha.12`

**Pull request merged**